Steer/DigitizerWorkflow/src/CTPDigitizerSpec.cxx

```diff
@@ -0,0 +1,132 @@
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/DataProcessorSpec.h"
+#include "Framework//Task.h"
+#include "DetectorsBase/BaseDPLDigitizer.h"
+#include "DataFormatsCTP/Digits.h"
+#include "Steer/HitProcessingManager.h" // for DigitizationContext
+#include "DetectorsCommonDataFormats/DetID.h"
+#include "CTPSimulation/Digitizer.h"
+#include "DataFormatsFT0/Digit.h"
+#include "DataFormatsFV0/BCData.h"
+
+#include <TStopwatch.h>
+
+using namespace o2::framework;
+namespace o2
+{
+namespace ctp
+{
+class CTPDPLDigitizerTask : public o2::base::BaseDPLDigitizer
+{
+  using GRP = o2::parameters::GRPObject;
+
+ public:
+  CTPDPLDigitizerTask() : o2::base::BaseDPLDigitizer(), mDigitizer() {}
+  ~CTPDPLDigitizerTask() override = default;
+  void initDigitizerTask(framework::InitContext& ic) override
+  {
+    mDigitizer.init();
+  }
+  void run(framework::ProcessingContext& pc)
+  {
+    if (mFinished) {
+      return;
+    }
+    // read collision context from input
+    //auto context = pc.inputs().get<o2::steer::DigitizationContext*>("collisioncontext");
+    //const bool withQED = context->isQEDProvided();
+    //auto& timesview = context->getEventRecords(withQED);
+    // read ctp inputs from input
+    auto ft0inputs = pc.inputs().get<std::vector<o2::ft0::DetTrigInput>>("ft0");
+    auto fv0inputs = pc.inputs().get<std::vector<o2::fv0::DetTrigInput>>("fv0");
+
+    // if there is nothing to do ... return
+    if ((ft0inputs.size() == 0) && (fv0inputs.size() == 0)) {
+      return;
+    }
+    std::map<o2::InteractionRecord, o2::ctp::CTPDigit> finputs;
+    TStopwatch timer;
+    timer.Start();
+    LOG(INFO) << "CALLING CTP DIGITIZATION";
+    //for (const auto& coll : timesview) {
+    //  mDigitizer.setInteractionRecord(coll);
+    //  mDigitizer.process(mDigits);
+    //  mDigitizer.flush(mDigits);
+    //}
+    for (const auto& inp : ft0inputs) {
+      CTPInputDigit finpdigit;
+      finpdigit.mDetector = o2::detectors::DetID::FT0;
+      finpdigit.mInputsMask = inp.mInputs;
+      CTPDigit fctpdigit;
+      fctpdigit.mIntRecord = inp.mIntRecord;
+      fctpdigit.mInputs.push_back(finpdigit);
+      finputs[inp.mIntRecord] = fctpdigit;
+    }
+    for (const auto& inp : fv0inputs) {
+      CTPInputDigit finpdigit;
+      finpdigit.mDetector = o2::detectors::DetID::FV0;
+      finpdigit.mInputsMask = inp.mInputs;
+      if (finputs.count(inp.mIntRecord) == 0) {
+        CTPDigit fctpdigit;
+        fctpdigit.mIntRecord = inp.mIntRecord;
+        fctpdigit.mInputs.push_back(finpdigit);
+        finputs[inp.mIntRecord] = fctpdigit;
+      } else {
+        finputs[inp.mIntRecord].mInputs.push_back(finpdigit);
+      }
+    }
+    for (const auto& inps : finputs) {
+      mDigitizer.setInteractionRecord(inps.first);
+      mDigitizer.process(inps.second, mDigits);
+      mDigitizer.flush(mDigits);
+    }
+    // send out to next stage
+    pc.outputs().snapshot(Output{"CTP", "DIGITS", 0, Lifetime::Timeframe}, mDigits);
```

Comment:
You should also send your readout (triggered) mode to GRPUpdater, it expects this info for all detectors, see e.g. https://github.com/AliceO2Group/AliceO2/blob/8ea4ea2797b756494b070bec8f6a7904e9381dff/Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx#L158-L159 (use `o2::parameters::GRPObject::PRESENT`, like TRD, which is a flag for detectors receiving external input)

Reply:
done